File: flutter_module/lib/models/cpu_info.dart
import 'dart:math';

class CPUInfo {
  late int maxFrequency;
  late int minFrequency;
  late int currMaxFrequency;
  late int currMinFrequency;
  late double maxPercentage;
  late double minPercentage;

  /// This is used to calculate current percentage
  late int _diffFrequency;

  /// For example, 4 x 1.90 GHz | 4 x 2.46 GHZ
  late String cpuInfo;
  bool hasData = false;

  CPUInfo(Map<String, Object?>? json) {
    hasData = json != null;

    // Unpack data from map
    cpuInfo = (json?['info'] as String?) ?? 'Unknown';

    maxFrequency = (json?['max'] as int?) ?? 0;
    minFrequency = (json?['min'] as int?) ?? 0;

    currMaxFrequency = (json?['max_curr'] as int?) ?? 0;
    currMinFrequency = (json?['min_curr'] as int?) ?? 0;

    // NOTE: Make sure it is not 0 to prevent zero division
    _diffFrequency = max(maxFrequency - minFrequency, 1);

    maxPercentage = calcCurrentPercent(true);
    minPercentage = calcCurrentPercent(false);
  }

  double calcCurrentPercent(bool max) {
    if (max) {
      return (currMaxFrequency - minFrequency) / _diffFrequency;
    } else {
      return (currMinFrequency - minFrequency) / _diffFrequency;
    }
  }

  int calcFrequency(double percent) {
    final offset = (_diffFrequency * percent).toInt();
    return offset + minFrequency;
  }

  /// Make sure max is always more than min and min is always less than max
  void onChangeSlider(double value, {required bool max}) {
    if (max) {
      maxPercentage = value;
      currMaxFrequency = calcFrequency(value);
      if (currMaxFrequency < currMinFrequency) {
        currMinFrequency = currMaxFrequency;
        minPercentage = maxPercentage;
      }
    } else {
      minPercentage = value;
      currMinFrequency = calcFrequency(value);
      if (currMinFrequency > currMaxFrequency) {
        currMaxFrequency = currMinFrequency;
        maxPercentage = minPercentage;
      }
    }
  }
}

File: cpuspeed/app/src/main/java/com/yihengquan/cpuspeed/data/CPUInfo.kt
package com.yihengquan.cpuspeed.data

data class CoreFrequency(
    val core: Int,
    val maxFreq: Int = 0,
    val minFreq: Int = 0,
    val curFreq: Int = 0,
)

data class CPUInfo(
    val maxFrequency: Int = 0,
    val minFrequency: Int = 0,
    val currMaxFrequency: Int = 0,
    val currMinFrequency: Int = 0,
    val cpuInfo: String = "Unknown",
    val coreCount: Int = 1,
    val hasData: Boolean = false,
    val coreFrequencies: List<CoreFrequency> = emptyList(),
) {
    val diffFrequency: Int get() = maxOf(maxFrequency - minFrequency, 1)

    val maxPercent: Float
        get() = if (hasData) (currMaxFrequency - minFrequency).toFloat() / diffFrequency else 0f

    val minPercent: Float
        get() = if (hasData) (currMinFrequency - minFrequency).toFloat() / diffFrequency else 0f

    fun calcFrequency(percent: Float): Int {
        return (percent * diffFrequency + minFrequency).toInt()
    }

    fun calcCoreFrequency(percent: Float, coreMin: Int, coreMax: Int): Int {
        val diff = maxOf(coreMax - coreMin, 1)
        return (percent * diff + coreMin).toInt()
    }
}
